File: server/languages/csharp/source/UnitTestCodegen.cs
using System;
using System.Text;
using System.Collections.Generic;

namespace language_csharp
{
	public class UnitTestCodegen
	{
		public static byte[] performCodegen(CodegenManager codegen, string source, IList<UnitTest> tests)
		{
			//Declare the UnitTestResults class so that it can then be serialised
			StringBuilder headerCode = new StringBuilder();
			headerCode.Append(@"
			class TestCaseResult {
				public string output; public string type; public bool exception;
				public TestCaseResult(string output, string type, bool exception) { this.output = output; this.type = type; this.exception = exception; }
			}
			class UnitTestResults {
				public System.Collections.Generic.List<TestCaseResult> result = new System.Collections.Generic.List<TestCaseResult>();
			}
			");
			headerCode.Append(source);
			
			//Generate the unit test boilerplate to populate the body of the main() method
			StringBuilder bodyCode = new StringBuilder();
			bodyCode.Append(@"
			var origStdout = System.Console.Out;
			var origStderr = System.Console.Error;
			System.Console.SetOut(System.IO.TextWriter.Null);
			System.Console.SetError(System.IO.TextWriter.Null);
			var results = new System.Collections.Generic.List<UnitTestResults>();
			UnitTestResults currTest = null;
			");
			
			//Generate the code for each of the unit tests
			foreach (UnitTest test in tests)
			{
				bodyCode.Append("currTest = new UnitTestResults();\n");
				bodyCode.Append(test.Setup);
				bodyCode.Append("\n");
				foreach (TestCase testCase in test.Cases)
				{
					bodyCode.Append(testCase.Setup);
					bodyCode.Append("\n");
					bodyCode.Append("try { var lastResult = " + test.Invocation + "(" + String.Join(",", testCase.Inputs) + ");\n");
					bodyCode.Append("currTest.result.Add(new TestCaseResult(lastResult.ToString(), lastResult.GetType().Name, false)); }\n");
					bodyCode.Append("catch (System.Exception e) { currTest.result.Add(new TestCaseResult(e.Message, e.GetType().Name, true)); }\n");
					bodyCode.Append(testCase.Teardown);
					bodyCode.Append("\n");
				}
				bodyCode.Append("results.Add(currTest);\n");
				bodyCode.Append(test.Teardown);
				bodyCode.Append("\n");
			}
			
			//Complete the boilerplate code
			bodyCode.Append(@"
			System.Console.SetOut(origStdout);
			System.Console.SetError(origStderr);
			System.Console.WriteLine(Newtonsoft.Json.JsonConvert.SerializeObject(results));
			");
			
			//Perform codegen
			return codegen.performCodegen(headerCode.ToString(), bodyCode.ToString());
		}
	}
}

File: server/languages/cxx/source/UnitTestCodegen.cpp
#include "UnitTestCodegen.h"

#include <boost/algorithm/string/join.hpp>
#include <sstream>

string UnitTestCodegen::performCodegen(CodegenManager* codegen, const string& source, const vector<UnitTest>& tests)
{
	//Generate the top-level code that follows the supplied source code
	std::stringstream headerCode;
	headerCode << source;
	headerCode << R"(

#include <json/json.h>
#include <iostream>
#include <sstream>
#include <string>
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include <typeinfo>
#include <cxxabi.h>

template <typename T>
std::string toString(const T& i)
{
	std::stringstream buf;
	buf << i;
	return buf.str();
}

std::string demangle(const std::string& mangled)
{
	int status = 0;
	char* result = abi::__cxa_demangle(mangled.c_str(), nullptr, nullptr, &status);
	if (status == 0)
	{
		std::string demangled(result);
		free(result);
		return demangled;
	}
	
	//Demangling failed
	return mangled;
}

)";
	
	//Generate the unit test boilerplate to populate the body of the main() method
	std::stringstream mainCode;
	mainCode << R"(
	Json::Value results(Json::arrayValue);
	Json::Value currResult;
	Json::Value currCase;
	FILE* devNull = fopen("/dev/null", "w");
	int origStdout = dup(fileno(stdout));
	int origStderr = dup(fileno(stderr));
	dup2(fileno(devNull), fileno(stdout));
	dup2(fileno(devNull), fileno(stderr));
	)";
	
	//Generate the code for each of the unit tests
	for (auto test : tests)
	{
		mainCode << "currResult = Json::Value(Json::objectValue);\n\t";
		mainCode << "currResult[\"result\"] = Json::Value(Json::arrayValue);\n\t";
		mainCode << test.setup() << "\n";
		for (int i = 0; i < test.cases_size(); ++i)
		{
			auto testCase = test.cases(i);
			auto inputs = testCase.inputs();
			mainCode << testCase.setup() << "\n";
			mainCode << "currCase = Json::Value(Json::objectValue);\n";
			mainCode << "try { auto lastResult = " << test.invocation() << "(" << boost::algorithm::join(inputs, ",") << ");\n";
			mainCode << "currCase[\"output\"] = toString(lastResult);\n";
			mainCode << "currCase[\"type\"] = demangle(typeid(lastResult).name());\n";
			mainCode << "currCase[\"exception\"] = false;\n";
			mainCode << "}";
			mainCode << "catch (std::runtime_error& e) {\n";
			mainCode << "currCase[\"output\"] = e.what();\n";
			mainCode << "currCase[\"type\"] = demangle(typeid(e).name());\n";
			mainCode << "currCase[\"exception\"] = true;\n";
			mainCode << "}";
			mainCode << "currResult[\"result\"].append(currCase);\n";
			mainCode << testCase.teardown() << "\n";
		}
		mainCode << "results.append(currResult);\n\t";
		mainCode << test.teardown() << "\n";
	}
	
	//Complete the boilerplate code
	mainCode << "dup2(origStdout, fileno(stdout));\n\t";
	mainCode << "dup2(origStderr, fileno(stderr));\n\t";
	mainCode << "std::cout << results.toStyledString() << std::endl;\n\t";
	
	//Perform codegen
	return codegen->performCodegen(headerCode.str(), mainCode.str());
}
